src/mc-fle2-find-equality-payload-v2.c

```diff
@@ -0,0 +1,151 @@
+                     bson_iter_type (&iter));                              \
+         goto fail;                                                        \
+      }                                                                    \
+      bson_iter_binary (&iter, &subtype, &len, &data);                     \
+      if (subtype != BSON_SUBTYPE_BINARY) {                                \
+         CLIENT_ERR ("Field '" #Name                                       \
+                     "' expected to be bindata subtype %d, got: %d",       \
+                     BSON_SUBTYPE_BINARY,                                  \
+                     subtype);                                             \
+         goto fail;                                                        \
+      }                                                                    \
+      if (!_mongocrypt_buffer_copy_from_binary_iter (&out->Dest, &iter)) { \
+         CLIENT_ERR ("Unable to create mongocrypt buffer for BSON binary " \
+                     "field in '" #Name "'");                              \
+         goto fail;                                                        \
+      }                                                                    \
+   }                                                                       \
+   END_IF_FIELD
+
+#define CHECK_HAS(Name)                                    \
+   if (!has_##Name) {                                      \
+      CLIENT_ERR ("Missing field '" #Name "' in payload"); \
+      goto fail;                                           \
+   }
+
+bool
+mc_FLE2FindEqualityPayloadV2_parse (mc_FLE2FindEqualityPayloadV2_t *out,
+                                  const bson_t *in,
+                                  mongocrypt_status_t *status)
+{
+   bson_iter_t iter;
+   bool has_d = false, has_s = false, has_l = false, has_cm = false;
+
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT_PARAM (in);
+
+   mc_FLE2FindEqualityPayloadV2_init (out);
+   if (!bson_validate (in, BSON_VALIDATE_NONE, NULL) ||
+       !bson_iter_init (&iter, in)) {
+      CLIENT_ERR ("invalid BSON");
+      return false;
+   }
+
+   while (bson_iter_next (&iter)) {
+      const char *field = bson_iter_key (&iter);
+      BSON_ASSERT (field);
+
+      PARSE_BINARY (d, edcDerivedToken)
+      PARSE_BINARY (s, escDerivedToken)
+      PARSE_BINARY (l, serverDerivedFromDataToken)
+      IF_FIELD (cm)
+      {
+         if (!BSON_ITER_HOLDS_INT64 (&iter)) {
+            CLIENT_ERR ("Field 'cm' expected to hold an int64");
+            goto fail;
+         }
+         out->maxContentionCounter = bson_iter_int64 (&iter);
+      }
+      END_IF_FIELD
+   }
+
+   CHECK_HAS (d);
+   CHECK_HAS (s);
+   CHECK_HAS (l);
+   CHECK_HAS (cm);
+
+   return true;
+fail:
+   mc_FLE2FindEqualityPayloadV2_cleanup (out);
+   return false;
+}
+
+#define IUPS_APPEND_BINDATA(name, subtype, value)              \
```

Comment:
```suggestion
#define APPEND_BINDATA(name, subtype, value)              \
```

I think IUPS stood for "Insert Update Payload Serialize". Suggest removing the prefix.